File: app/src/main/java/igrek/todotree/commands/PersistenceCommand.java
package igrek.todotree.commands;


import java.io.File;
import java.io.IOException;

import javax.inject.Inject;

import igrek.todotree.dagger.DaggerIOC;
import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.exceptions.DeserializationFailedException;
import igrek.todotree.logger.Logger;
import igrek.todotree.service.backup.Backup;
import igrek.todotree.service.backup.BackupManager;
import igrek.todotree.service.filesystem.FilesystemService;
import igrek.todotree.service.filesystem.PathBuilder;
import igrek.todotree.service.history.ChangesHistory;
import igrek.todotree.service.preferences.Preferences;
import igrek.todotree.service.preferences.PropertyDefinition;
import igrek.todotree.service.resources.UserInfoService;
import igrek.todotree.service.tree.TreeManager;
import igrek.todotree.service.tree.TreeScrollCache;
import igrek.todotree.service.tree.persistence.TreePersistenceService;
import igrek.todotree.ui.contextmenu.BackupListMenu;

public class PersistenceCommand {
	
	@Inject
	TreeManager treeManager;
	
	@Inject
	UserInfoService userInfo;
	
	@Inject
	BackupManager backupManager;
	
	@Inject
	TreeScrollCache scrollCache;
	
	@Inject
	FilesystemService filesystem;
	
	@Inject
	Preferences preferences;
	
	@Inject
	TreePersistenceService persistenceService;
	
	@Inject
	ChangesHistory changesHistory;
	
	@Inject
	Logger logger;
	
	public PersistenceCommand() {
		DaggerIOC.getFactoryComponent().inject(this);
	}
	
	void optionReload() {
		treeManager.reset();
		scrollCache.clear();
		loadRootTree();
		new GUICommand().updateItemsList();
		userInfo.showInfo("Database loaded.");
	}
	
	void optionSave() {
		saveDatabase();
		userInfo.showInfo("Database saved.");
	}
	
	void saveDatabase() {
		if (!changesHistory.hasChanges()) {
			logger.info("No changes have been made - skipping saving");
			return;
		}
		
		saveRootTree();
		backupManager.saveBackupFile();
	}
	
	public void loadRootTree() {
		filesystem.mkdirIfNotExist(filesystem.externalSDPath().toString());
		String dbFilePath = preferences.getValue(PropertyDefinition.dbFilePath, String.class);
		if (dbFilePath.startsWith("/")) { //absolute
			loadDbFromFile(dbFilePath);
		} else { // relative
			PathBuilder dbFilePathBuilder = filesystem.externalSDPath().append(dbFilePath);
			loadDbFromFile(dbFilePathBuilder.toString());
		}
	}
	
	public void loadRootTreeFromBackup(Backup backup) {
		PathBuilder dbFilePath = filesystem.externalSDPath()
				.append(preferences.getValue(PropertyDefinition.dbFilePath, String.class));
		String path = dbFilePath.parent().append(backup.getFilename()).toString();
		loadDbFromFile(path);
		changesHistory.registerChange();
	}
	
	private void loadDbFromFile(String dbFilePath) {
		changesHistory.clear();
		logger.info("Loading database from file: " + dbFilePath);
		if (!filesystem.exists(dbFilePath)) {
			userInfo.showInfo("Database file does not exist. Default empty database loaded.");
			return;
		}
		try {
			String fileContent = filesystem.openFileString(dbFilePath);
			// AbstractTreeItem rootItem = SimpleTreeSerializer.loadTree(fileContent); // porting db to JSON
			AbstractTreeItem rootItem = persistenceService.deserializeTree(fileContent);
			treeManager.setRootItem(rootItem);
			logger.info("Database loaded.");
		} catch (IOException | DeserializationFailedException e) {
			changesHistory.registerChange();
			logger.error(e);
			userInfo.showInfo("Failed to load database: " + e.getMessage());
		}
	}
	
	private void saveRootTree() {
		PathBuilder dbFilePath = filesystem.externalSDPath()
				.append(preferences.getValue(PropertyDefinition.dbFilePath, String.class));
		try {
			String output = persistenceService.serializeTree(treeManager.getRootItem());
			//Logger.debug("Serialized data: " + output);
			filesystem.saveFile(dbFilePath.toString(), output);
		} catch (IOException e) {
			logger.error(e);
		}
		logger.debug("Database saved successfully.");
	}
	
	public void optionRestoreBackup() {
		new BackupListMenu().show();
	}
}

File: app/src/main/java/igrek/todotree/intent/PersistenceCommand.kt
package igrek.todotree.intent

import igrek.todotree.exceptions.DeserializationFailedException
import igrek.todotree.info.UiInfoService
import igrek.todotree.info.errorcheck.SafeExecutor
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.service.backup.Backup
import igrek.todotree.service.backup.BackupManager
import igrek.todotree.service.filesystem.FilesystemService
import igrek.todotree.service.history.ChangesHistory
import igrek.todotree.service.tree.TreeManager
import igrek.todotree.service.tree.TreeScrollCache
import igrek.todotree.service.tree.persistence.TreePersistenceService
import igrek.todotree.ui.contextmenu.BackupListMenu
import java.io.File
import java.io.IOException

class PersistenceCommand(
    treeManager: LazyInject<TreeManager> = appFactory.treeManager,
    uiInfoService: LazyInject<UiInfoService> = appFactory.uiInfoService,
    backupManager: LazyInject<BackupManager> = appFactory.backupManager,
    treeScrollCache: LazyInject<TreeScrollCache> = appFactory.treeScrollCache,
    filesystemService: LazyInject<FilesystemService> = appFactory.filesystemService,
    treePersistenceService: LazyInject<TreePersistenceService> = appFactory.treePersistenceService,
    changesHistory: LazyInject<ChangesHistory> = appFactory.changesHistory,
) {
    private val treeManager by LazyExtractor(treeManager)
    private val uiInfoService by LazyExtractor(uiInfoService)
    private val backupManager by LazyExtractor(backupManager)
    private val treeScrollCache by LazyExtractor(treeScrollCache)
    private val filesystemService by LazyExtractor(filesystemService)
    private val treePersistenceService by LazyExtractor(treePersistenceService)
    private val changesHistory by LazyExtractor(changesHistory)
    private val settingsState by LazyExtractor(appFactory.settingsState)

    private val logger = LoggerFactory.logger

    fun optionReload() {
        SafeExecutor {
            treeManager.reset()
            treeScrollCache.clear()
            loadDbFromFile(dbFile())
            uiInfoService.showInfo("Database loaded.")
        }
        GUICommand().updateItemsList()
    }

    fun saveDatabaseUi() {
        val saved = saveDatabase()
        if (saved)
            uiInfoService.showInfo("Database saved.")
    }

    fun saveDatabase(): Boolean {
        if (!changesHistory.hasChanges()) {
            logger.info("No changes have been made - skipping saving")
            return false
        }
        saveRootTree()
        backupManager.saveBackupFile()
        changesHistory.clear()
        return true
    }

    fun loadRootTree() {
        SafeExecutor {
            loadDbFromFile(dbFile())
        }
    }

    private fun dbFile(): File {
        val appDataDir = filesystemService.appDataRootDir()
        return appDataDir.resolve("todo.json")
    }

    private fun dbCopyFile(): File {
        return filesystemService.internalStorageAppDir().resolve("todo.json")
    }

    private fun externalBackupFile(): File? {
        val path = settingsState.externalBackupPath.takeIf { it.isNotBlank() } ?: return null
        val dir = File(path)
        dir.mkdirs()
        return dir.resolve("todo.json")
    }

    fun loadRootTreeFromBackup(backup: Backup) {
        SafeExecutor {
            treeManager.reset()
            treeScrollCache.clear()
            val backupDir = filesystemService.appDataSubDir("backup")
            val path = backupDir.resolve(backup.filename)
            loadDbFromFile(path)
            changesHistory.registerChange()
            uiInfoService.showInfo("Database backup loaded: " + backup.filename)
        }
        GUICommand().updateItemsList()
    }

    fun loadRootTreeFromImportedFile(fileContent: String, filename: String) {
        SafeExecutor {
            treeManager.reset()
            treeScrollCache.clear()
            loadDbFromFileContent(fileContent)
            changesHistory.registerChange()
            uiInfoService.showInfo("Database imported from $filename")
        }
        GUICommand().updateItemsList()
    }

    private fun loadDbFromFile(dbFile: File) {
        changesHistory.clear()
        logger.debug("Loading database from file: $dbFile")
        if (!dbFile.exists()) {
            throw RuntimeException("Database file does not exist: $dbFile. Default empty database loaded.")
        }
        try {
            val fileContent = filesystemService.openFileString(dbFile.absolutePath)
            val rootItem = treePersistenceService.deserializeTree(fileContent)
            treeManager.rootItem = rootItem
            logger.info("Database loaded from file: $dbFile")
        } catch (e: IOException) {
            changesHistory.registerChange()
            throw RuntimeException("Failed to load database: " + e.message)
        } catch (e: DeserializationFailedException) {
            changesHistory.registerChange()
            throw RuntimeException("Failed to load database: " + e.message)
        }
    }

    private fun loadDbFromFileContent(fileContent: String) {
        changesHistory.clear()
        try {
            val rootItem = treePersistenceService.deserializeTree(fileContent)
            treeManager.rootItem = rootItem
        } catch (e: IOException) {
            changesHistory.registerChange()
            throw RuntimeException("Failed to load database: " + e.message)
        } catch (e: DeserializationFailedException) {
            changesHistory.registerChange()
            throw RuntimeException("Failed to load database: " + e.message)
        }
    }

    private fun saveRootTree() {
        try {
            val output = treePersistenceService.serializeTree(treeManager.rootItem)
            val dbFilePath = dbFile().absolutePath
            val internalBackupFilePath = dbCopyFile().absolutePath // backup copy in the internal storage
            val externalBackupFile = externalBackupFile() // backup copy in the external storage (optional)

            filesystemService.saveFile(dbFilePath, output)
            logger.info("Database saved to $dbFilePath")

            filesystemService.saveFile(internalBackupFilePath, output)
            logger.debug("Database copy saved to $internalBackupFilePath")

            if (externalBackupFile != null) {
                val externalBackupFilePath = externalBackupFile.absolutePath
                filesystemService.saveFile(externalBackupFilePath, output)
                logger.debug("Database copy saved to $externalBackupFilePath")
            }
        } catch (e: IOException) {
            logger.error(e)
        }
    }

    fun optionRestoreBackup() {
        BackupListMenu().show()
    }

}